specs/capri/Monitoring Design.md

```diff
@@ -0,0 +1,723 @@
+
+-   Ceph exporter available in Prometheus:
+    <https://github.com/digitalocean/ceph_exporter>
+
+-   Also got radosgw exporter:
+    <https://github.com/blemmenes/radosgw_usage_exporter>
+
+*A Detailed Diagram*
+
+![](media/be7e84d9043ecad275a20996ac0b855f.jpg)
+
+OpenSDS will provide Metrics REST APIs.
+
+The ML module of OpenSDS or a 3rd-party consumer of the OpenSDS Metrics module
+will make a POST request, asking for metrics, using OpenSDS Metrics REST API,
+possibly running some cron jobs. The request will be sent to the OpenSDS Metrics
+Controller, and then to the Dock.
+
+On the Dock there will be a Collector which collects metrics from Metric
+drivers.
+
+There will be a Metrics driver for every storage backend. We need to define
+Metrics driver interfaces based on what data to collect. Performance related
+data will be the main focus for Capri.
+
+After the Collector receives data from the Metric driver, the data gets
+processed by an Adapter. The Adapter component includes a Converter and an
+Exporter.
+
+1.  The Converter converts the data to a proper format that can be understood by
+    the receiving end. For example, the Prometheus Converter converts data to a
+    format that Prometheus can understand.
+
+2.  The Exporter sends(emits) the data to the intended destination.
+
+The following Adapters will be developed:
+
+1.  Adapter to emit data to Prometheus
+
+    1.  Including a converter and an exporter.
+
+2.  Adapter to send data to Kafka
+
+3.  3rdparty Adapter to send data to specified destination
+
+How often to generate data?
+
+1.  Metrics will be collected when OpenSDS Metrics API is called. (higher
+    priority than \#2)
+
+2.  OpenSDS will also collect data about each volume operation when it happens
+    and emits in Prometheus format so it will be gathered by the Prometheus
+    server.
+
+Note: OpenSDS does not store these data in its own database. It will be stored
+in Prometheus if the Prometheus adapter is used. Consumer of the metrics data,
+i.e. the OpenSDS ML module, will use OpenSDS Metrics API to retrieve data.
+OpenSDS controller will convert the request to a REST API call to Prometheus
+asking for the data.
+
+*Description*
+--
+
+*UI*
+
+1.  Integrate Grafana into the OpenSDS UI. Prometheus is configured as a data
+    source in Grafana, dashboards are defined and integrated into OpenSDS
+
+2.  The access control, authentication etc. are handled in the OpenSDS UI
+
+*API SERVER*
+
+1.  APIs include a metrics GET API and POST API, explained below
+
+2.  GET API - to get the collected metrics from Prometheus. This API request
+    will be passed to the controller to translate to a PromQL call, get the data
+    from Prometheus and return it to the caller.
+
+3.  POST API - to collect metrics from the driver. Typically, this will be
+    called by a 3rd party. The API server passes this call to the controller to
+    generate metrics from the metrics drivers.
+
+*CONTROLLER*
```

Comment:
If I get it right, api-server and controller only accesses Prometheus when receiving GET API, dock can access Prometheus when receiving both POST and GET API.

Reply:
GET API won’t come to dock. Controller will direct GET calls to Prometheus. For the POST call, you are right.